modules/nextflow/src/main/groovy/nextflow/extension/GroupTupleOp.groovy

```diff
@@ -55,20 +62,21 @@ class GroupTupleOp {
 
     private sort
 
-    GroupTupleOp(Map params, DataflowReadChannel source) {
+    private OpContext context = new ContextRunPerThread()
```

Comment:
Why does groupTuple use the run-per-thread context? Same question for collate. I would expect both to use ContextGrouping, whereas ContextRunPerThread seems to be for combining multiple input channels. But groupTuple and collate are both single-threaded.

Reply:
`ContextRunPerThread` is needed when an operator uses multiple `DataflowProcessor`s under the hood. Likely here also `ContextSequential` should work, because the real logic is provided by this snipper that unwrap the values and create a new `OperatorRun` instead with the corresponding inputs ids 

https://github.com/nextflow-io/nextflow/blob/61c153c01263443d08932562a41bae3699ae845b/modules/nextflow/src/main/groovy/nextflow/extension/GroupTupleOp.groovy#L168-L177

Reply:
I guess you could make it work with either the ContextGrouping or the ContextSequential. Still trying to understand how the OpDatum is being used but it seems like a good way to associate inputs with outputs when the mapping is not simply 1-to-1.

I figure we should avoid the ContextRunPerThread where it isn't needed, since both collate and groupTuple have only one DataflowProcessor

Reply:
OpDatum keep together a concrete value with the OperatorRun instance that acquired it and ultimately the set of inputs IDs. When the operator emission is composed, a new run instance is created with all inputs that concurred to that outputs. Makes sense?

Reply:
Yes that makes sense. Perhaps the better question is, what does an "operator run" mean for an operator like groupTuple? Each input is a separate "run" in one sense, but only one of those "runs" will emit an output. So is there an OperatorRun for each input or one for each emitted group?

I will write some tests and use my mermaid diagram to test some of these operators. I suspect something is missing for buffer / collate / groupTuple

Reply:
In principle it should be an instance for each run, but in the reality it's used as object holder to associate an output message with the upstream inputs. See here 

https://github.com/nextflow-io/nextflow/blob/0e30a8f0b50ba3e800631b6708f100cf2c7b9f33/modules/nextflow/src/main/groovy/nextflow/prov/Tracker.groovy#L114-L131

Reply:
Right. Maybe it would be better to rename `OperatorRun` to `OperatorLink`, and rename `TrailRun` to `ProvenanceLink`, to clarify their purpose as a link from a set of inputs to an output

Reply:
Nice. Changed `TrailRun` to `ProvLink`, keeping `OperatorRun` for now